storage/local/storage.go

```diff
@@ -1249,26 +1249,39 @@ func (s *MemorySeriesStorage) loop() {
 	memoryFingerprints := s.cycleThroughMemoryFingerprints()
 	archivedFingerprints := s.cycleThroughArchivedFingerprints()
 
+	// Checkpoints can happen concurrently with maintenance so even with heavy
+	// checkpointing there will still be sufficient progress on maintenance.
+	checkpointLoopStopped := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-s.loopStopping:
+				checkpointLoopStopped <- struct{}{}
+				return
+			case <-checkpointTimer.C:
+				err := s.persistence.checkpointSeriesMapAndHeads(s.fpToSeries, s.fpLocker)
+				if err != nil {
+					log.Errorln("Error while checkpointing:", err)
+				} else {
+					dirtySeriesCount = 0
```

Comment:
This is a race as it is now set in two different goroutines. (The other in L1287.) Should be easily fixable with the `atomic` package.

Also, since now series may become dirty _during_ checkpointing, we need to set this at the beginning of a checkpoint, not at the end. (This will then be an upper bound, not a precise number anymore, but better than underestimating it and then deal with numerous support cases about hour long crash recoveries again, as we did before the dirty-series feature was introduced.)